### PR TITLE
Integrate daterange into datasets API

### PR DESCRIPTION
**Pull request by dbutenhof**
PBENCH-1125

Remove the `GET /datasets/daterange` API in favor of a new query parameter, as `GET /datasets?daterange`. This allows getting the actual date range for any filtered selection of datasets.

### CODE REVIEW

File: lib/pbench/server/api/resources/datasets_list.py [hunk 1]
```diff
@@ -348,6 +349,27 @@ def keyspace(self, query: Query) -> JSONOBJECT:
...
+    def daterange(self, request: Request, json: JSONOBJECT, query: Query) -> JSONOBJECT:
+        """Return only the date range of the selected datasets.
+
+        Replace the selected "entities" (normally Dataset columns) with the
+        SQL min and max functions on the dataset upload timestamp so that the
+        generated SQL query will return a tuple of those two values.
+
+        Args:
+            request: The HTTP Request object
+            json: The JSON query parameters
```
Comment (webbnh) [inline lib/pbench/server/api/resources/datasets_list.py:361]: These two arguments are not referenced in the function...should we note that in some way?  (IIRC, you don't particularly like using `_` as a parameter name...perhaps prefix the parameter names with an underscore, or, at least, add an annotation to their argument descriptions?)

File: lib/pbench/server/api/resources/datasets_list.py [hunk 2]
```diff
@@ -497,5 +519,7 @@ def _get(
         query = self._build_sql_query(owner, json.get("access"), query)
         if json.get("keysummary"):
             return jsonify(self.keyspace(query))
+        elif json.get("daterange"):
+            return jsonify(self.daterange(request, json, query))
         else:
             return jsonify(self.datasets(request, json, query))
```
Comment (webbnh) [inline lib/pbench/server/api/resources/datasets_list.py:525]: Do we "like" having `"keysummary"` quietly override `"daterange"`?

The obvious alternatives being:
- since we return a JSON object anyway, we could return both the key summary and the date range, if they are both requested; or,
- we could return an error if both `"keysummary"` and `"daterange"` are specified.

The documentation doesn't describe what is supposed to happen if both are specified.  (BTW, in the documentation, the `keysummary` section is not in alphabetical order, like the rest of the keys.)

File: lib/pbench/test/unit/server/test_datasets_list.py [hunk 3]
```diff
@@ -698,3 +698,71 @@ def test_key_summary(self, query_as):
...
+        Args:
+            name_list: List of dataset names
+
+        Returns:
+            {"from": first_date, "to": last_date}
+        """
+        from_time = datetime.datetime.now(datetime.timezone.utc)
+        to_time = datetime.datetime(
+            year=1970, month=1, day=1, tzinfo=datetime.timezone.utc
+        )
```
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_list.py:718]: You're sure that "now" is far enough in the future?  I would go with something like:
```python
        from_time = datetime.datetime.fromtimestamp(
            9999999999, tz=datetime.timezone.utc
	)
        to_time = datetime.datetimefromtimestamp(0, tz=datetime.timezone.utc)
```

File: lib/pbench/test/unit/server/test_datasets_list.py [hunk 4]
```diff
@@ -698,3 +698,71 @@ def test_key_summary(self, query_as):
...
+        for name in sorted(name_list):
+            dataset = Dataset.query(name=name)
+            to_time = max(dataset.uploaded, to_time)
+            from_time = min(dataset.uploaded, from_time)
```
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_list.py:722]: Alternatively, you could do something like this:
```python
        to_time = max(Dataset.query(name=n).uploaded for n in name_list)
        from_time = min(Dataset.query(name=n).uploaded for n in name_list)
```
This does do two database queries, but in the context of the unit tests that's probably not an issue.  On the other hand, it saves you from having to initialize `to_time` and `from_time` to crazy values.  (BTW, I don't see any reason why we need to sort `name_list` before aggregating it...do you?)

However, it has a nasty failure mode when `name_list` is empty...but, then, your proposed code has an "interesting" behavior in that case, too....  🙂

**Review by webbnh — CHANGES_REQUESTED** [review]
Nice improvement!!

However, I think the docs (and possibly the code) need a tweak around the behavior when _both_ `keysummary` and `daterange` are specified.  And, I have some suggestions/nits.

**Comment by webbnh** [comment]
And, we need to get @npalaska to fix those flakey unit tests!  😉

File: lib/pbench/server/api/resources/datasets_list.py (hunk 1 above)
Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_list.py:361]: I don't like using `_` because you lose the documentation on how the method is *called* without searching up the call tree. Yes, I perhaps unnecessarily kept all three of the helpers consistent, but I like consistency and it doesn't present any real issues. (And, e.g., I could have created a dispatcher binding `helper` to one of the three and then make a single call: if there were four instead of three, I would have, and I considered it briefly at three but decided not to.)

I don't see how it's helpful to try to document that they're "not used" as this means nothing unless we're trying to encourage callers to `None` them out, which I don't like. Again, this is interface documentation and we're documenting the parameters, not how they're currently used internally.

File: lib/pbench/server/api/resources/datasets_list.py (hunk 2 above)
Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_list.py:525]: I wouldn't so much say that I "like" it as that I didn't see a reason to care enough to worry about it. For one thing, I'd really need to demote the keysummary output to a sub-object, to combine them, and while I spent whole nanoseconds carefully weighing this idea during coding, I decided to be lazy and not mess with more tests. E.g., something like `{"from": "<date>", "to": "<date>", "keys": {"dataset": {...}}}`

Then again, maybe that's not such a bad idea since the UI could make a single "setup" call to get the key namespace and date range on page load instead of needing to make separate calls...

File: lib/pbench/test/unit/server/test_datasets_list.py (hunk 3 above)
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_datasets_list.py:718]: I literally lifted this helper from the old `test_datasets_daterange.py` and the only change was adding the `_daterange` in the name since we already had a `get_results`. I don't see any point in worrying about the implementation now. If this were production code, yeah, "now" isn't late enough to catch time travelling performance engineers, but for our unit test environment there's no advantage and it's not worth the churn.

File: lib/pbench/test/unit/server/test_datasets_list.py (hunk 4 above)
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_datasets_list.py:722]: And, um, it turns out that the query itself has a problem in that case, so I *do* need to deal with the empty list case, on both ends. And the `sorted` is, yeah, unnecessary here, but was simply copied originally from the datasets list test and apparently didn't seem worth changing. Huh.

File: lib/pbench/server/api/resources/datasets_list.py (hunk 1 above)
Comment (webbnh) [reply lib/pbench/server/api/resources/datasets_list.py:361]: It helps ease the dissonance in the reader/reviewer when they see the parameters in the function signature and find that they aren't used in the function.  I figured that that you were trying to keep the interface consistent with that of other functions, which is why I didn't suggest removing them, but it would be good to have some acknowledgement that, _in the particular case of **this** function_, those parameters are not used, _because_ they are part of a common interface.

Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_list.py:361]: In fact, I noticed that only one of the other helpers used the whole set, so they're already inconsistent and there's no point in passing the additional parameters. Oops. I just cloned the last helper signature without really thinking about it. 😟

File: lib/pbench/server/api/resources/datasets_list.py (hunk 2 above)
Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_list.py:525]: So, I'll allow them to be combined. Why not?

File: docs/API/V1/list.md [hunk 5]
```diff
@@ -252,83 +259,93 @@ will return a JSON document with the keys `config`, `date`, `hostname_f`,
...
 }
 ```
+
+### Combining key namespace summary and date range
+
+When both the `keysummary` and `daterange` query parameters are `true`, the
+`application/json` response body contains the `from`, `to`, and `keys` key
+values. If the selected collection filters produce no results, as with
+`daterange` alone, the `from` and `to` keys will be omitted and the value of
+`keys` will be an empty object.
```
Comment (webbnh) [inline docs/API/V1/list.md:351]: Should we have verbiage in the "Key namespace summary" section, like we do in the "Dataset date range" section, which describes what is returned when no results are selected?

Er, but...doesn't [this line](https://github.com/distributed-system-analysis/pbench/blob/4e79e67fcc9cf4386d693aff1301c67a7f8d182a/lib/pbench/server/api/resources/datasets_list.py#L336) prevent the `keys` object from ever being empty?

(Either way, we should probably have a unit test which checks the documentation assertion here....)

File: lib/pbench/test/unit/server/test_datasets_list.py [hunk 6]
```diff
@@ -728,6 +730,7 @@ def get_daterange_results(
             ("drb", {"owner": "drb"}, ["drb", "fio_1"]),
             ("drb", {"mine": "true"}, ["drb", "fio_1"]),
             ("drb", {"access": "public"}, ["fio_1", "fio_2"]),
+            ("drb", {"name": "noname"}, []),
```
Comment (webbnh) [inline lib/pbench/test/unit/server/test_datasets_list.py:733]: Do we also have a test case where the result is empty due to the choice of dates?  (If not...should we?)

**Review by webbnh — DISMISSED** [review]
Looks good, but I have an awkward question (or two) which probably means that we're missing a unit test.

File: docs/API/V1/list.md (hunk 5 above)
Comment (dbutenhof) [reply docs/API/V1/list.md:351]: Yeah, you're right; and thinking about it now I'm not sure how I want to resolve that. We're always describing the `dataset` SQL table columns, which, arguably, we "shouldn't" be doing if there are no matches. However testing whether there are matches is a little messy since you can't really `if <iterator>` or `if len(<iterator>)`, and I'd rather not list-ify it... but the alternative is to do it on the first iteration of the `for`, which is (almost?) equally ugly if a bit more scaleable.

**Review by webbnh — DISMISSED** [review]
Looks good but for a typo.  Also, I have a couple of suggestions.

File: lib/pbench/server/api/resources/datasets_list.py [hunk 7]
```diff
@@ -331,22 +332,45 @@ def keyspace(self, query: Query) -> JSONOBJECT:
...
+                )
             for m in d.metadatas:
                 # "metalog" is a top-level key in the Metadata schema, but we
                 # report it as a sub-key of "dataset".
                 if m.key == Metadata.METALOG:
                     self.accumulate(aggregate["dataset"], m.key, m.value)
                 else:
                     self.accumulate(aggregate, m.key, m.value)
-        return aggregate
+        return {"keys": aggregate}
```
[the remark this answers is not in the document]
Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_list.py:351]: > I can't help tinkering with this (having to do the not aggregate check for every dataset is dissatisfying...). Perhaps this will suit?

Ugh. So we're going to compare two dicts to see whether to throw it away? The "is-empty" check is relatively simple, and I don't think *this* is an improvement.

(It wouldn't work anyway without doing a *deep* copy of `base_aggregate` since otherwise you're just aliasing the collection and they'll always be the same; and the deep copy is yet another unnecessary expense.)

> It also occurs to me that, since we're constructing an object to return at the end anyway, perhaps we can move the other `if` statement outside the loop:

We'd have to skip the `metalog` key in the first loop, or else waste time computing it and then `del aggregate["metalog"]` at the end. We don't want both `metalog` and `dataset.metalog`. I'd rather keep it as-is.

> And, if we can assume that `datasets` is non-empty iff `Metadata.METALOG in aggregate` is true, then we can combine the ternary in the `return` statement the `if Metadata.METALOG` statement. (Is that correct?)

I'm not even sure what you mean here; but ultimately I'd rather leave it alone!

File: lib/pbench/server/api/resources/datasets_list.py [hunk 8]
```diff
@@ -495,7 +519,20 @@ def _get(
...
-            return jsonify(self.keyspace(query))
-        else:
-            return jsonify(self.datasets(request, json, query))
+            result.update(self.keyspace(query))
+            done = True
+        if json.get("daterange"):
+            result.update(self.daterange(query))
+            done = True
+        if not done:
+            result = self.datasets(request, json, query)
```
[the remark this answers is not in the document]
Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_list.py:537]: It bothered me to repeat the key names. Yeah, I considered it, but it felt even more ugly.

I need an `if x: andif y: else:` construct here (the `else` fires only if neither of the non-exclusive chained `if` conditions fires) 😆

**Review by dbutenhof — COMMENTED** [review]
I'll fix the doc typo ...

File: lib/pbench/server/api/resources/datasets_list.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (webbnh) [reply lib/pbench/server/api/resources/datasets_list.py:351]: > we're going to compare two dicts to see whether to throw it away? The "is-empty" check is relatively simple, and I don't think _this_ is an improvement.

The dict-dict comparison fails quickly if they are different, and, if they are not different, then the two dicts are small and straightforward to compare -- so either way it's a limited expense; on the other side of the balance is the "is-empty" check which is a very small incremental cost but it will build up if we have a large number of datasets

> It wouldn't work anyway without doing a _deep_ copy of `base_aggregate` since otherwise you're just aliasing the collection and they'll always be the same

That's a good point.  However, instead of capturing `base_aggregate`, we could just compare `aggregate` to the original comprehension again.

> and the deep copy is yet another unnecessary expense

As I said, it think the initial aggregation value is small, and I expect that it would be easy to regenerate; the question is, how expensive is the alternative (e.g., what if there are lots of datasets?).

> We'd have to skip the `metalog` key in the first loop, or else waste time computing it and then `del aggregate["metalog"]` at the end. We don't want both `metalog` and `dataset.metalog`.

I don't follow.  I'm proposing that we generate the subtree for `metalog` where it naturally falls, and then, when it's all done, move it to where we want it.  Adding it as `dataset.metalog` is a cheap operation (it's just a reference copy), and then removing the top-level `metalog` key is cheap as well.  (It's not quite as _pretty_ as I would like, but it's not expensive.)  I don't think we have to skip anything, and the cost of moving at the end is offset by not having to check each key for `"metalog"`.

> > And, if we can assume that `datasets` is non-empty iff `Metadata.METALOG in aggregate` is true, then we can combine the ternary in the `return` statement the `if Metadata.METALOG` statement. (Is that correct?)
> 
> I'm not even sure what you mean here; but ultimately I'd rather leave it alone!

I was wondering whether we could detect an empty datasets list by finding no `metalog` key in the aggregate, but I don't know whether that approach hangs together.

But, since you would prefer to leave it alone, I shall resolve this conversation.  🙂

**Review by webbnh — APPROVED** [review]
👍

**Review by riya-17 — APPROVED** [review]
